tests/test_backoff.py

```diff
@@ -0,0 +1,322 @@
+    def test_make_request_max_retries_reached_for_Server5xxError(
+        self, mock_get_timezone, mock_refresh_token
+    ):
+        """
+        Test that make_request raises Server5xxError after exceeding maximum retry attempts.
+        """
+        mock_get_timezone.return_value = "UTC"
+        mock_refresh_token.return_value = "dummy_refresh_token"
+        client = CampaignMonitorClient(self.config)
+
+        with patch("requests.request") as mock_request:
+            error_response = MagicMock()
+            error_response.status_code = 500
+
+            mock_request.side_effect = [error_response] * 6
+
+            with self.assertRaises(Server5xxError):
+                client.make_request("https://dummy-url.com", "GET")
+            self.assertEqual(mock_request.call_count, 5)
+
+    @patch("tap_campaign_monitor.client.CampaignMonitorClient.refresh_access_token")
+    @patch("tap_campaign_monitor.client.CampaignMonitorClient.get_timezone")
+    def test_make_request_max_retries_reached_for_Server429Error(
+        self, mock_get_timezone, mock_refresh_token
+    ):
+        """
+        Test that make_request raises Server429Error after exceeding maximum retry attempts.
+        """
+        mock_get_timezone.return_value = "UTC"
+        mock_refresh_token.return_value = "dummy_refresh_token"
+        client = CampaignMonitorClient(self.config)
+
+        with patch("requests.request") as mock_request:
+            error_response = MagicMock()
+            error_response.status_code = 429
+
+            mock_request.side_effect = [error_response] * 6
+
+            with self.assertRaises(Server429Error):
+                client.make_request("https://dummy-url.com", "GET")
+            self.assertEqual(mock_request.call_count, 5)
+
+    @patch("tap_campaign_monitor.client.CampaignMonitorClient.refresh_access_token")
+    @patch("tap_campaign_monitor.client.CampaignMonitorClient.get_timezone")
+    def test_make_request_429_with_valid_retry_after(
+        self, mock_get_timezone, mock_refresh_token
+    ):
+        """
+        Test that make_request respects the retry_after value from X-RateLimit-Reset header
+        when handling Server429Error.
+        """
+        mock_get_timezone.return_value = "UTC"
+        mock_refresh_token.return_value = "dummy_refresh_token"
+        client = CampaignMonitorClient(self.config)
+
+        with patch("requests.request") as mock_request:
+            error_response = MagicMock()
+            error_response.status_code = 429
+            error_response.headers = {"X-RateLimit-Reset": "10"}
+
+            mock_request.side_effect = [error_response] * 2 + [
+                MagicMock(status_code=200, json=lambda: {"data": "ok"})
+            ]
+
+            result = client.make_request("https://dummy-url.com", "GET")
+            self.assertEqual(result, {"data": "ok"})
+            self.assertEqual(mock_request.call_count, 3)
```

Comment:
We need to verify whether `_rate_limit_backoff()` is actually getting called and what value it returns.